lxd/storage/drivers/driver_btrfs_utils.go

```diff
@@ -0,0 +1,547 @@
+package drivers
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/lxc/lxd/lxd/util"
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/ioprogress"
+	"github.com/lxc/lxd/shared/logger"
+	"golang.org/x/sys/unix"
+)
+
+var errBtrfsNoQuota = fmt.Errorf("Quotas disabled on filesystem")
+var errBtrfsNoQGroup = fmt.Errorf("Unable to find quota group")
+
+func btrfsSubvolumeCreate(subvol string) error {
+	parentDestPath := filepath.Dir(subvol)
+	if !shared.PathExists(parentDestPath) {
+		err := os.MkdirAll(parentDestPath, 0711)
+		if err != nil {
+			return err
+		}
+	}
+
+	_, err := shared.RunCommand(
+		"btrfs",
+		"subvolume",
+		"create",
+		subvol)
+	if err != nil {
+		logger.Errorf("Failed to create BTRFS subvolume \"%s\": %v", subvol, err)
+		return err
+	}
+
+	return nil
+}
+
+func btrfsSubvolumesGet(path string) ([]string, error) {
+	result := []string{}
+
+	if !strings.HasSuffix(path, "/") {
+		path = path + "/"
+	}
+
+	// Unprivileged users can't get to fs internals
+	filepath.Walk(path, func(fpath string, fi os.FileInfo, err error) error {
+		// Skip walk errors
+		if err != nil {
+			return nil
+		}
+
+		// Ignore the base path
+		if strings.TrimRight(fpath, "/") == strings.TrimRight(path, "/") {
+			return nil
+		}
+
+		// Subvolumes can only be directories
+		if !fi.IsDir() {
+			return nil
+		}
+
+		// Check if a btrfs subvolume
+		if isBtrfsSubvolume(fpath) {
+			result = append(result, strings.TrimPrefix(fpath, path))
+		}
+
+		return nil
+	})
+
+	return result, nil
+}
+
+func btrfsSubvolumeSnapshotsGet(path string) ([]string, error) {
+	result := []string{}
+
+	if !strings.HasSuffix(path, "/") {
+		path = path + "/"
+	}
+
+	// Unprivileged users can't get to fs internals
+	filepath.Walk(path, func(fpath string, fi os.FileInfo, err error) error {
+		// Skip walk errors
+		if err != nil {
+			return nil
+		}
+
+		// Ignore the base path
+		if strings.TrimRight(fpath, "/") == strings.TrimRight(path, "/") {
+			return nil
+		}
+
+		// Subvolumes can only be directories
+		if !fi.IsDir() {
+			return nil
+		}
+
+		// Check if a btrfs subvolume snapshot
+		if isBtrfsSubvolume(fpath) {
+			result = append(result, strings.TrimPrefix(fpath, path))
+		}
+
+		return nil
+	})
+
+	return result, nil
+}
+
+// isBtrfsSubvolume returns true if the given Path is a btrfs subvolume else
+// false.
+func isBtrfsSubvolume(subvolPath string) bool {
+	fs := unix.Stat_t{}
+	err := unix.Lstat(subvolPath, &fs)
+	if err != nil {
+		return false
+	}
+
+	// Check if BTRFS_FIRST_FREE_OBJECTID
+	if fs.Ino != 256 {
+		return false
+	}
+
+	return true
+}
+
+func isOnBtrfs(path string) bool {
```

Comment:
All function names in this file should be prefixed by `btrfs`, maybe rename to `btrfsIsValid` or something along those lines?

Reply:
Actually this could be replaced by a common function in `utils.go` to get the filesystem type from a path, then the driver can just compare it with the right magic.